### PR TITLE
Create rand function for Base.KeySet and Base.ValueIterator{Dict}

### LINKED ISSUE #51605: Making `rand` for a `Base.KeySet` faster
Consider this simple benchmark

```julia
using BenchmarkTools, Random

d = Dict(x => x for x in 1:1000);
rng = Xoshiro(42)

@benchmark rand($rng, keys($d))
```

on my computer under 1.9.3, it returns 

```julia
julia> @benchmark rand($rng, keys($d))
BenchmarkTools.Trial: 10000 samples with 10 evaluations.
 Range (min … max):  510.000 ns …  43.100 μs  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):       1.530 μs               ┊ GC (median):    0.00%
 Time  (mean ± σ):     1.565 μs ± 562.739 ns  ┊ GC (mean ± σ):  0.00% ± 0.00%

                   ▁▂▄▂▅▇▇▇█▇▅█▇▅▄▅ ▁▁
  ▁▁▁▁▁▁▁▂▂▂▃▄▄▃▆▆▇███████████████████▇▆▅▄▃▃▃▂▂▂▂▁▂▁▁▁▁▁▁▁▁▁▁▁▁ ▄
  510 ns           Histogram: frequency by time         2.86 μs <

 Memory estimate: 0 bytes, allocs estimate: 0.
```

It sounded a bit slow (and more the number of elements gets bigger more it is so) then I tried

```julia
function Base.rand(rng::AbstractRNG, keys_d::Base.KeySet) 
    return Base.rand(rng, keys_d.dict).first
end
@benchmark rand($rng, keys($d))
```

which gave me

```julia
julia> @benchmark rand($rng, keys($d))
BenchmarkTools.Trial: 10000 samples with 994 evaluations.
 Range (min … max):  34.004 ns … 113.380 ns  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     36.217 ns               ┊ GC (median):    0.00%
 Time  (mean ± σ):   36.632 ns ±   3.040 ns  ┊ GC (mean ± σ):  0.00% ± 0.00%

        ▃▂█▃▂▂
  ▂▂▂▃▅▅██████▅▄▃▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▁▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▁▂▂ ▃
  34 ns           Histogram: frequency by time         48.8 ns <

 Memory estimate: 0 bytes, allocs estimate: 0.
```

Can't something similar be used instead?

In general I think that the current method iterates over the `KeySet` for finding a random value, while I think that the new one requires almost just a constant cost

### PR DESCRIPTION
**Pull request by ameligrana**
Fixes #51605 . 

I also created a version of rand for `Base.ValueIterator{Dict}` which currently throws an error.

### CODE REVIEW

**Comment by oscardssmith** [comment]
Does this work as is or do you also need to define the `Sampler` constructor? (not rhetorical, I actually don't know the answer)

**Comment by Tortar** [comment]
Tested with Revise before doing the PR, it should work as it is

File: stdlib/Random/src/generation.jl [hunk 1]
```diff
@@ -452,6 +452,12 @@ function rand(rng::AbstractRNG, sp::SamplerSimple{<:Dict,<:Sampler})
     end
 end
 
+rand(rng::AbstractRNG, sp::SamplerTrivial{<:Base.KeySet{<:T,<:Dict}} where T) =
```
Comment (rfourquet) [inline stdlib/Random/src/generation.jl:455]:
```suggestion
rand(rng::AbstractRNG, sp::SamplerTrivial{<:Base.KeySet{<:Any,<:Dict}}) =
```

**Review by rfourquet — COMMENTED** [review]
LGTM, but this would need tests which exercise the new code paths.

**Comment by Tortar** [comment]
tests added and locally run

**Pull request merged**